crossbar/worker/container.py

```diff
@@ -243,6 +243,47 @@ def panic(fail, msg):
                     session.disconnect()
                 session._swallow_error = panic
 
+                #   This is for RAW sockets where the onClose handler is in protocol session
+                #   rather than the protocol itself.
+                #
+                def close_wrapper(original_handler, was_clean):
+                    """
+                    Handle component shutdown (onClose)
+                    """
+                    return_value = original_handler(was_clean)
+                    #
+                    #   Make sure we're still open for business ...
+                    #
+                    if component_id not in self.components:
+                        self.log.warn('Component already closed ({})'.format(component_id))
+                        return return_value
+                    #
+                    #   Report the exit type
+                    #
+                    if was_clean:
+                        self.log.info('Component shutdown Ok ({})'.format(component_id))
+                    else:
+                        self.log.error('Component shutdown Errored ({})'.format(component_id))
+                    #
+                    #   Tidy up .. self.components is cleaned up in "stop_component"
+                    #
+                    component = self.components[component_id]
+                    self._publish_component_stop(component)
+                    component._stopped.callback(component.marshal())
+                    #
+                    #   Report and exit ...
+                    #
+                    self.log.info('Container hosting ({}) components, exit mode is ({})'.format(
+                                  len(self.components),
+                                  self._exit_mode))
+                    if not self.components and self._exit_mode == self.SHUTDOWN_ON_LAST_COMPONENT_STOPPED:
+                        self.log.info('Stopping Container')
+                        self.shutdown()
+                    return return_value
+                #
+                #   Assign onClose handler
+                #
+                session.onClose = partial(close_wrapper, session.onClose)
                 return session
 
             except Exception:
```

Comment:
There's a LOT of code in this "try/except Exception" block; can we at least move the inner-functions out? How is the new close_wrapper here for raw-sockets only? (i.e. I don't see any code that actually checks what "session" is..?)

Reply:
Yes, you're right, it should be more granular, let me reorganise ..